File: ios/Sources/ServerAddress.swift
import Foundation

/// Turns what someone actually types into a usable server URL.
///
/// Volunteers set this once, on a phone, in a dark booth. Demanding a
/// well-formed URL is a bad trade when the intent is nearly always obvious:
///
///     172.16.1.41            -> http://172.16.1.41:7883
///     172.16.1.41:7883       -> http://172.16.1.41:7883
///     comms.church.org       -> https://comms.church.org
///     HTTPS://Comms.Church.Org/  -> https://comms.church.org
///
/// The scheme is inferred rather than assumed: a private or link-local
/// address is a bench setup with no certificate, so it gets http and the
/// default port. Anything else is a real hostname behind Caddy, so it gets
/// https on 443. An explicit scheme or port always wins.
enum ServerAddress {
    static let defaultLocalPort = 7883

    static func normalize(_ raw: String) -> URL? {
        var text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return nil }

        // Drop any path someone pasted along with the host — we always talk
        // to /token, and a stray /token in the setting would double up.
        var scheme: String?
        if let range = text.range(of: "://") {
            scheme = String(text[text.startIndex ..< range.lowerBound]).lowercased()
            text = String(text[range.upperBound...])
        }
        if let slash = text.firstIndex(of: "/") {
            text = String(text[text.startIndex ..< slash])
        }
        guard !text.isEmpty else { return nil }

        // Split host:port, being careful not to mangle a bracketed IPv6 host.
        var host = text
        var port: Int?
        if !text.hasPrefix("[") {
            let parts = text.split(separator: ":", maxSplits: 1, omittingEmptySubsequences: false)
            if parts.count == 2, let parsed = Int(parts[1]), (1 ... 65535).contains(parsed) {
                host = String(parts[0])
                port = parsed
            } else if parts.count == 2 {
                return nil // "host:notaport"
            }
        }

        host = host.lowercased()
        guard !host.isEmpty, !host.contains(" ") else { return nil }

        let local = isLocal(host)
        let resolvedScheme = scheme ?? (local ? "http" : "https")
        guard resolvedScheme == "http" || resolvedScheme == "https" else { return nil }

        var components = URLComponents()
        components.scheme = resolvedScheme
        components.host = host
        // Only default the port for a bench setup. A real hostname is behind
        // Caddy on 443, where naming the port would be wrong.
        if let port {
            components.port = port
        } else if local, resolvedScheme == "http" {
            components.port = defaultLocalPort
        }

        return components.url
    }

    /// Private, loopback, link-local, or mDNS — anything that cannot have a
    /// public certificate.
    static func isLocal(_ host: String) -> Bool {
        if host == "localhost" || host.hasSuffix(".local") { return true }

        let octets = host.split(separator: ".").compactMap { Int($0) }
        guard octets.count == 4, octets.allSatisfy({ (0 ... 255).contains($0) }) else { return false }

        switch (octets[0], octets[1]) {
        case (10, _): return true
        case (127, _): return true
        case (192, 168): return true
        case (169, 254): return true
        case (172, 16 ... 31): return true
        default: return false
        }
    }
}

File: web/src/server-address.js
// Turns what someone actually types into a usable server URL.
//
// Mirrors ios/Sources/ServerAddress.swift deliberately — the two clients must
// agree on what "192.168.1.50" means, or the same note taped to the wall gives
// different results depending on which phone you hold.
//
//   172.16.1.41            -> http://172.16.1.41:7883
//   comms.church.org       -> https://comms.church.org
//   HTTPS://Comms.Church.Org/  -> https://comms.church.org

export const DEFAULT_LOCAL_PORT = 7883;

/** Private, loopback, link-local or mDNS — anything that cannot have a public cert. */
export function isLocal(host) {
  if (host === "localhost" || host.endsWith(".local")) return true;

  const octets = host.split(".").map((part) => (/^\d+$/.test(part) ? Number(part) : NaN));
  if (octets.length !== 4 || octets.some((n) => !Number.isInteger(n) || n < 0 || n > 255)) {
    return false;
  }

  const [a, b] = octets;
  if (a === 10 || a === 127) return true;
  if (a === 192 && b === 168) return true;
  if (a === 169 && b === 254) return true;
  if (a === 172 && b >= 16 && b <= 31) return true;
  return false;
}

/** @returns {string|null} a normalised absolute URL, or null if it makes no sense. */
export function normalize(raw) {
  let text = String(raw ?? "").trim();
  if (!text) return null;

  let scheme = null;
  const schemeAt = text.indexOf("://");
  if (schemeAt !== -1) {
    scheme = text.slice(0, schemeAt).toLowerCase();
    text = text.slice(schemeAt + 3);
  }

  // Drop any path pasted along with the host; we always talk to /token, and a
  // stray /token in the setting would double up.
  const slash = text.indexOf("/");
  if (slash !== -1) text = text.slice(0, slash);
  if (!text) return null;

  let host = text;
  let port = null;
  if (!text.startsWith("[")) {
    const colon = text.indexOf(":");
    if (colon !== -1) {
      const tail = text.slice(colon + 1);
      if (!/^\d+$/.test(tail)) return null;
      const parsed = Number(tail);
      if (parsed < 1 || parsed > 65535) return null;
      host = text.slice(0, colon);
      port = parsed;
    }
  }

  host = host.toLowerCase();
  if (!host || host.includes(" ")) return null;

  const local = isLocal(host);
  const resolvedScheme = scheme ?? (local ? "http" : "https");
  if (resolvedScheme !== "http" && resolvedScheme !== "https") return null;

  // Only default the port for a bench setup; a real hostname sits behind Caddy
  // on 443, where naming the port would be wrong.
  let suffix = "";
  if (port !== null) suffix = `:${port}`;
  else if (local && resolvedScheme === "http") suffix = `:${DEFAULT_LOCAL_PORT}`;

  return `${resolvedScheme}://${host}${suffix}`;
}
